### DISCUSSION TITLE
esptool woes

### DISCUSSION THREAD

**Discussion by tve** [Moddable-OpenSource/moddable#1143] [General]
I'm having a hell of a time getting mcconfig to load firmware onto a moddable two using my linux box. It fails most of the time. Is this a known issue? (I've been using esp32's for years and understand the process.)
The problem is that esptool fails to put the chip into programming mode. Typical outcome:
```
> /home/tve/.espressif/python_env/idf4.4_py3.10_env/bin/python /home/sw/moddable/esp-idf/components/esptool_py/esptool/esptool.py --chip esp32 -p /d
ev/ttyUSB1 -b 460800 --before=default_reset chip_id
esptool.py v3.3.2
Serial port /dev/ttyUSB1
Connecting......................................

A fatal error occurred: Failed to connect to ESP32: Wrong boot mode detected (0x13)! The chip needs to be in download mode.
```
On a whim I used my "std" installed esptool:
```
> esptool.py --chip esp32 -p /dev/ttyUSB1 -b 460800 --before=default_reset chip_id
esptool.py v4.5.1
Serial port /dev/ttyUSB1
Connecting.....
Chip is ESP32-D0WD (revision v1.0)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: fc:f5:c4:61:a2:f8
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 460800
Changed.
Warning: ESP32 has no Chip ID. Reading MAC instead.
MAC: fc:f5:c4:61:a2:f8
Hard resetting via RTS pin...
```
Note the esptool version difference... Are others experiencing this type of issue as well or am I a lucky guy?
To use the installed esptool version, is my best choice to hack make.esp32.mk?

**Comment by phoddie** [comment]
This is somewhat outside the Moddable SDK. It isn't our tool.

But... jumping into programming mode is timing sensitive, so there is a bit of luck. And Espressif seems to tweak that code with some regularity.

The Moddable SDK uses the esptool.py from the ESP-IDF:

```
which esptool.py
~/esp32/esp-idf/components/esptool_py/esptool/esptool.py
```

That is currently v3.3.2 in the ESP-IDF version we target.

You do seem to have a newer version. And that works for you. That's great.
